tests/0113-cooperative_rebalance.cpp

```diff
@@ -257,79 +369,1698 @@ void run_test(std::string &t1, std::string &t2,
+
+    expect_rebalance("rejoin group", c,
+                     RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS,
+                     rd_false/*expect lost*/, 10+2);
+
+    TEST_SAY("Closing consumer\n");
+    test_consumer_close(c);
+
+    TEST_SAY("Destroying consumer\n");
+    rd_kafka_destroy(c);
+
+    TEST_SAY("Destroying mock cluster\n");
+    test_mock_cluster_destroy(mcluster);
+  }
+
+
+  /* ------- r_lost_partitions_commit_illegal_generation_test
+   *
+   * check lost partitions revoke occurs on ILLEGAL_GENERATION Commit
+   * error.
+   */
+
+  static void r_lost_partitions_commit_illegal_generation_test () {
+    TEST_SAY("Executing r_lost_partitions_commit_illegal_generation_test\n");
+
+    const char *bootstraps;
+    rd_kafka_mock_cluster_t *mcluster;
+    const char *groupid = "mygroup";
+    const char *topic = "test";
+    rd_kafka_t *c;
+    rd_kafka_conf_t *conf;
+
+    mcluster = test_mock_cluster_new(3, &bootstraps);
+
+    rd_kafka_mock_coordinator_set(mcluster, "group", groupid, 1);
+
+    /* Seed the topic with messages */
+    test_produce_msgs_easy_v(topic, 0, 0, 0, 100, 10,
+                             "bootstrap.servers", bootstraps,
+                             "batch.num.messages", "10",
+                             NULL);
+
+    test_conf_init(&conf, NULL, 30);
+    test_conf_set(conf, "bootstrap.servers", bootstraps);
+    test_conf_set(conf, "security.protocol", "PLAINTEXT");
+    test_conf_set(conf, "group.id", groupid);
+    test_conf_set(conf, "auto.offset.reset", "earliest");
+    test_conf_set(conf, "enable.auto.commit", "false");
+    test_conf_set(conf, "partition.assignment.strategy", "cooperative-sticky");
+
+    c = test_create_consumer(groupid, rebalance_cb, conf, NULL);
+
+    test_consumer_subscribe(c, topic);
+
+    expect_rebalance("initial assignment", c,
+                     RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS,
+                     rd_false/*don't expect lost*/, 5+2);
+
+    /* fail heartbeats */
+    rd_kafka_mock_push_request_errors(
+      mcluster, RD_KAFKAP_OffsetCommit,
+      5,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION,
+      RD_KAFKA_RESP_ERR_ILLEGAL_GENERATION);
+
+    rd_kafka_commit(c, NULL, rd_false);
+
+    expect_rebalance("lost partitions", c,
+                     RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS,
+                     rd_true/*expect lost*/, 10+2);
+
+    expect_rebalance("rejoin group", c,
+                     RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS,
+                     rd_false/*expect lost*/, 20+2);
+
+    TEST_SAY("Closing consumer\n");
+    test_consumer_close(c);
+
+    TEST_SAY("Destroying consumer\n");
+    rd_kafka_destroy(c);
+
+    TEST_SAY("Destroying mock cluster\n");
+    test_mock_cluster_destroy(mcluster);
+  }
+
+
+  int main_0113_cooperative_rebalance (int argc, char **argv) {
+    int i;
+
```

Comment:
Identify a set of fundamental/important tests and then leave out the rest if test_quick is set.

Reply:
ooh, that's hard. ok, I'll choose some.